api/nodes/views.py

```diff
@@ -99,15 +149,191 @@ def get_queryset(self):
+    given your permissions on the object.
+
+    NOTE: Most of the API will be stable as far as how the links work because the things they are accessing are fairly
+    stable and predictable, so if you felt the need, you could construct them in the normal REST way and they should
+    be fine.
+    The 'self' links from the NodeFilesList may have to change from time to time, so you are highly encouraged to use
+    the links as we provide them before you use them, and not to reverse engineer the structure of the links as they
+    are at any given time.
+    """
+    serializer_class = NodeFilesSerializer
+
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        ContributorOrPublic,
+    )
+
+    def get_valid_self_link_methods(self, root_folder=False):
+        valid_methods = {'file': ['GET'], 'folder': [], }
+        user = self.request.user
+        if user is None or user.is_anonymous():
+            return valid_methods
+
+        permissions = self.get_node().get_permissions(user)
+        if 'write' in permissions:
+            valid_methods['file'].append('POST')
+            valid_methods['file'].append('DELETE')
+            valid_methods['folder'].append('POST')
+            if not root_folder:
+                valid_methods['folder'].append('DELETE')
+
+        return valid_methods
+
+    def get_file_item(self, item, cookie, obj_args):
+        file_item = {
+            'valid_self_link_methods': self.get_valid_self_link_methods()[item['kind']],
+            'provider': item['provider'],
+            'path': item['path'],
+            'name': item['name'],
+            'node_id': self.get_node()._primary_key,
+            'cookie': cookie,
+            'args': obj_args,
+            'waterbutler_type': 'file',
+            'item_type': item['kind'],
+        }
+        if file_item['item_type'] == 'folder':
+            file_item['metadata'] = {}
+        else:
+            file_item['metadata'] = {
+                'content_type': item['contentType'],
+                'modified': item['modified'],
+                'size': item['size'],
+                'extra': item['extra'],
+            }
+        return file_item
+
+    def get_queryset(self):
+        query_params = self.request.query_params
+
+        addons = self.get_node().get_addons()
+        user = self.request.user
+        cookie = None if self.request.user.is_anonymous() else user.get_or_create_cookie()
+        node_id = self.get_node()._id
+        obj_args = self.request.parser_context['args']
+
+        provider = query_params['provider'] if 'provider' in query_params else None
+        path = query_params['path'] if 'path' in query_params else '/'
+        files = []
+
+        if provider is None:
+            valid_self_link_methods = self.get_valid_self_link_methods(True)
+            for addon in addons:
+                if addon.config.has_hgrid_files:
```

Comment:
I know ```addon.config.has_hgrid_files``` is a relic of the past, but perhaps this PR is the place to rename this to something like: ```addon.config.has_files```